### DISCUSSION TITLE
Pass more data to scripts (script/shell extension)

### DISCUSSION THREAD

**Discussion by phorcys420** [espanso/espanso#1858] [Feature requests and Ideas]
Currently, espanso only passes environment variables for the config path and the output of other variables.

I think this should be further extended by adding the following variables :
| Variable name                     | Description                                                                                                                                               |
|-----------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------|
| `$ESPANSO_TRIGGER`                | Contains the name of the trigger that was used to call the script, adds the ability to modulate behavior based on trigger name. (e.g when using aliases). |
| `$ESPANSO_WINDOW_NAME`            | Window name, adds the ability to modulate behavior based on the current Window.                                                                           |
| `$ESPANSO_WINDOW_CLASS`           | Window class, espanso already has this data so you might as well pass it for future-proofing.                                                             |
| `$ESPANSO_WINDOW_EXECUTABLE_PATH` | Window executable path, same as the last one                                                                                                              |

In my case, `$ESPANSO_TRIGGER` would be especially useful as I implemented a workaround in PowerShell for #1201 that has to delete the trigger because espanso doesn't do it (since it waits for the script to exit).

For `$ESPANSO_WINDOW_NAME`, I have a script that looks at the contents of my clipboard to search into my company's password manager, I think this could be enhanced by putting a specific password related to an application at the top of my choice window.

---

NOTE: This seems relatively easy to implement, I will probably send a PR if I get the time.

**Comment by smeech** [comment]
There are workarounds for each of these, but they're fairly tortuous.
`$ESPANSO_TRIGGER` would be useful for regex triggers (plain triggers are already known but would need `echo:` variables) and avoid nested Named Capture Groups `(?P<> )`s to derive them.
The others can be found from within scripts, but often by loading additional libraries, which add to the overhead, increases the time for the script to return and risk failure of the expansion because the typist has moved on.
If the information is easy to extract within Espanso the process would be smoother as long as the additional overhead for each trigger is light. It occurs to me that, ideally, their generation should be omitted for non Shell/Script expansions., where they won't be of use (unless someone can think of a use-case for them).

**Reply by phorcys420** [reply]
> It occurs to me that, ideally, their generation should be omitted for non Shell/Script expansions., where they won't be of use (unless someone can think of a use-case for them).

yes, that is what I thought too, I don't know what the current behavior is for the `CONFIG` variable but I'll just copy that if I implement it to be consistent with what espanso already does.

**Reply by phorcys420** [reply]
currently it's only done in the `script` and `shell` extensions.
what strikes me is that they're handled individually, I think that it should maybe be unified before to avoid redundancy. (we can probably modify `convert_to_env_variables` to do just that)

https://github.com/espanso/espanso/blob/bd239906deffc1702d6dcb4390e8bfaef310cac2/espanso-render/src/extension/script.rs#L87-L91
https://github.com/espanso/espanso/blob/bd239906deffc1702d6dcb4390e8bfaef310cac2/espanso-render/src/extension/shell.rs#L221-L225

**Reply by phorcys420** [reply]
it seems that we should pass an [`AppInfo`](https://github.com/espanso/espanso/blob/bd239906deffc1702d6dcb4390e8bfaef310cac2/espanso-info/src/lib.rs#L42-L46) to the `calculate()` methods of this file.

or we could directly inject variables into the `scope`, but that seems kinda dirty

**Comment by smeech** [comment]
It would be useful to add $HOME and the Windows & macOS equivalents to this list, so that they're available to expansions which don't use a shell.

**Reply by phorcys420** [reply]
I don't think that would do what you intend, those lists are for the `script` and `shell` extension and I saw your discussion in the discord, I think it's something separate!
